proxy/http2/Http2ConnectionState.cc

```diff
@@ -1601,7 +1595,20 @@ Http2ConnectionState::send_goaway_frame(Http2StreamId id, Http2ErrorCode ec)
   SCOPED_MUTEX_LOCK(lock, this->ua_session->mutex, this_ethread());
   this->ua_session->handleEvent(HTTP2_SESSION_EVENT_XMIT, &frame);
 
-  handleEvent(HTTP2_SESSION_EVENT_FINI, nullptr);
+  this->ua_session->set_half_close_flag(true);
+  DebugHttp2Con(ua_session, "session half-close");
+
+  // As for gracefull shutdown, TS should process outstanding stream as long as possible.
+  // As for signal connection error, TS should close connection immediately.
+  if (ec != Http2ErrorCode::HTTP2_ERROR_NO_ERROR) {
```

Comment:
It depends on when you send second GOAWAY frame.

a)
Do you want to close connection immediately, after sending second GOAWAY frame? If so it make sense. 
1. send GOAWAY (INT32_MAX)
2. process outstanding streams
3. send GOAWAY (last stream id)
4. close connection

b) 
But if you want to hold connection for a while, you need to set timeout or schedule do_io_close in somewhere.
1. send GOAWAY (INT32_MAX)
2. send GOAWAY (last stream id)
3. process outstanding streams
4. close connection

IMO, second choice looks better. Because client can retry streams earlier.